### PR TITLE
Introduces new automations via Actions for managing incident-based discussions

### PR DESCRIPTION
**Pull request by alidacodes**
This PR introduces a set of workflows that allow us to automatically generate and update new discussions whenever there is a public incident declared within GitHub. This then allows us to centralize discussion without requiring manual intervention from either the incident commander or a community manager. This PR also extends the `discussion.rb` lib file to support the new workflows. 

## Demo & example screenshot

https://github.com/user-attachments/assets/5e1a3e16-d1c2-481b-b359-d4b5c5bac219

image

## Implementation overview

The workflows are meant to work as follows:

### `open-incident-discussion.yml`

This workflow acts upon the `incident-declared` repository dispatch event. It takes the public title and incident URL from the repository dispatch event and generates the new discussion via the `open_incident_discussion` action, which will be posted in the Announcements category. 

### `update-incident-discussion.yml`

This workflow acts upon the `incident-updated` repository dispatch event. It uses the URL slug to identify the correct discussion to update, then posts the update message to the relevant discussion as a comment. 

### `post-incident-summary.yml`

This workflow acts upon the `incident-public-summary` repository dispatch event. It again uses the URL slug to identify the correct discussion to update, then uses the `post_incident_summary` action to add a comment (like the update action) and to mark this new comment as the answer. It also updates the initial post body to include a graphic denoting that this incident has been resolved.

### `close-incident-discussions.yml`

This workflow is different from the rest, as it instead operates on a schedule to satisfy the requirement that the incident discussions are closed ~48 hours after they are opened. It runs two actions; the first, `check_open_incident_discussions`, which confirms if there are any eligible discussions for closure, and then runs the `close_incident_discussions` action to close them, ensuring that any discussions that haven't been marked as answered are first marked as answered by finding the last comment by the automation, marking that as the answer, and then closing the discussion.

## Considerations & notes

- There will be a fast-follow PR with the asset URL of the banner image once it has been uploaded to this repo
- These workflows will not respond to incidents until we update the configuration on GitHub's end to do so, which can be done once this PR has been approved and merged. This will include installing the corresponding GitHub app, `incident-discussion-bot`, on this repo; this app is used to connect the eventing on GitHub to the workflows in this repo.
- This PR will be merged once both engineering and community approves, and both are ready for this feature to be live.
- The final piece of this puzzle is Slack alerting, which can be configured directly after this has been merged. 

/cc @queenofcorgis

### CODE REVIEW

File: .github/actions/post_incident_summary.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+summary = "### Incident Summary \n #{ENV["INCIDENT_PUBLIC_SUMMARY"]}"
+comment_id = discussion.add_comment(body: summary).dig("data", "addDiscussionComment", "comment", "id")
+
+# mark this new comment as the answer
+# (note that we don't need the discussion's context for this, so we don't call it on the instance of Discussion but on the struct)
+Discussion.mark_comment_as_answer(comment_id:)
+
+# update the post body to include the resolved picture
+updated_body = "![A dark background with two security-themed abstract shapes positioned in the top left and bottom right corners. In the center of the image, bold white text reads \\\"Incident Resolved\\\" with a white Octocat logo.](https://github.com/community/incident-discussion-bot/blob/main/.github/src/incident_resolved.png?raw=true) \n \n #{selected_incident["body"]}"
+Discussion.update_discussion(body: updated_body)
```
Comment (alidacodes) [inline .github/actions/post_incident_summary.rb:24]: This is a class method and not an instance method still, as we need to be able to access it from the class level in https://github.com/community/community/pull/145114/files#diff-47c2ce836e71f40dd553801cbf090ce3311b07bda5aa9127268eb519106b0173R31

**Comment by sarasaran281972** [comment]
[sara_git.zip](https://github.com/user-attachments/files/17850160/sara_git.zip)
in the attachment ; The ip address is  = 172.31.31.10 Homer server running in cloude local ip address.
The IP address is 	117.246.168.249 my mobile phone hotspots address.( service provider address).

1 i have changed 👍 Edit the pg_hba.conf file in the same directory as postgresql.conf.
Add the following lines to allow remote connections:
host    all             all             0.0.0.0/0               md5
host    all             all             ::/0                    md5

Find the listen_addresses setting and change it to '*' to allow connections from any IP address:
listen_addresses = '*'

Ensure the port setting is set to the desired port number (default is 5432):
port = 5432

Problem:
if I try to access HOMER using the brower , using admin or postgres loging and credential i could not able to see the conent in the homer, kindly let me know is there is any installation I missed and like  give idea to resolve this issue.

**Review by mgriffin — COMMENTED** [review]
A few general comments to hopefully explain what I was talking about when I mentioned returning a `Discussion` object instead of a hash.

Feel free to go with whichever way you feel is better, they both work in the same way in the end

File: .github/lib/discussions.rb [hunk 2]
```diff
@@ -8,7 +8,10 @@
   :id,
   :url,
   :title,
-  :labelled
+  :labelled,
+  :body,
+  :created_at,
+  :is_answered
 ) do
```
Comment (alidacodes) [inline .github/lib/discussions.rb:15]: by adding `:is_answered` to the struct, I eliminate the need for an `is_answered?` lookup, reducing the overall calls needed for the `close_incident_discussions` action

File: .github/actions/open_incident_discussion.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+body = <<~BODY
+## :exclamation: An incident has been declared:
+
+**#{ENV['PUBLIC_TITLE']}**
+
+_Subscribe to this Discussion for updates on this incident. Please upvote or emoji react instead of commenting +1 on the Discussion to avoid overwhelming the thread. Any account guidance specific to this incident will be shared in thread and on the [Incident Status Page](#{ENV['INCIDENT_URL']})._
+BODY
+
+# we need to create a new discussion in the correct category with the correct label
```
Comment (mgriffin) [inline .github/actions/open_incident_discussion.rb:27]: Just for my own knowledge, the announcements are only being created in one category for now, is that right?

File: .github/workflows/close-incident-discussions.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+name: Close incident discussion after time elapsed
+
+on:
+  workflow_dispatch:
+  schedule:
+    # daily at midnight
```
Comment (mgriffin) [inline .github/workflows/close-incident-discussions.yml:6]: UTC? Or does Actions have a default timezone?

**Review by mgriffin — DISMISSED** [review]
😍 I love this!

File: .github/workflows/close-incident-discussions.yml (hunk 4 above)
Comment (alidacodes) [reply .github/workflows/close-incident-discussions.yml:6]: UTC, I can update the comment

File: .github/actions/open_incident_discussion.rb (hunk 3 above)
Comment (alidacodes) [reply .github/actions/open_incident_discussion.rb:27]: That's correct! They can be moved afterwards, but at this time we're only creating them in the Announcements category

**Review by mgriffin — DISMISSED** [review]

File: Gemfile.lock [hunk 5]
```diff
@@ -17,18 +17,14 @@ GEM
     concurrent-ruby (1.2.3)
     connection_pool (2.4.1)
     drb (2.2.1)
-    faraday (2.9.0)
-      faraday-net_http (>= 2.0, < 3.2)
-    faraday-net_http (3.1.0)
-      net-http
+    faraday (2.0.0)
```
Comment (mgriffin) [inline Gemfile.lock]: Faraday is being downgraded from 2.9.0 to 2.0.0, is that on purpose?

**Review by mgriffin — APPROVED** [review]

**Comment by wal497** [comment]
> This PR introduces a set of workflows that allow us to automatically generate and update new discussions whenever there is a public incident declared within GitHub. This then allows us to centralize discussion without requiring manual intervention from either the incident commander or a community manager. This PR also extends the `discussion.rb` lib file to support the new workflows.
> 
> ## Demo & example screenshot
>  Demo.Incident.Discussion.Automation.mp4 
> image
> 
> ## Implementation overview
> The workflows are meant to work as follows:
> 
> ### `open-incident-discussion.yml`
> This workflow acts upon the `incident-declared` repository dispatch event. It takes the public title and incident URL from the repository dispatch event and generates the new discussion via the `open_incident_discussion` action, which will be posted in the Announcements category.
> 
> ### `update-incident-discussion.yml`
> This workflow acts upon the `incident-updated` repository dispatch event. It uses the URL slug to identify the correct discussion to update, then posts the update message to the relevant discussion as a comment.
> 
> ### `post-incident-summary.yml`
> This workflow acts upon the `incident-public-summary` repository dispatch event. It again uses the URL slug to identify the correct discussion to update, then uses the `post_incident_summary` action to add a comment (like the update action) and to mark this new comment as the answer. It also updates the initial post body to include a graphic denoting that this incident has been resolved.
> 
> ### `close-incident-discussions.yml`
> This workflow is different from the rest, as it instead operates on a schedule to satisfy the requirement that the incident discussions are closed ~48 hours after they are opened. It runs two actions; the first, `check_open_incident_discussions`, which confirms if there are any eligible discussions for closure, and then runs the `close_incident_discussions` action to close them, ensuring that any discussions that haven't been marked as answered are first marked as answered by finding the last comment by the automation, marking that as the answer, and then closing the discussion.
> 
> ## Considerations & notes
> * There will be a fast-follow PR with the asset URL of the banner image once it has been uploaded to this repo
> * These workflows will not respond to incidents until we update the configuration on GitHub's end to do so, which can be done once this PR has been approved and merged. This will include installing the corresponding GitHub app, `incident-discussion-bot`, on this repo; this app is used to connect the eventing on GitHub to the workflows in this repo.
> * This PR will be merged once both engineering and community approves, and both are ready for this feature to be live.
> * The final piece of this puzzle is Slack alerting, which can be configured directly after this has been merged.
> 
> /cc @queenofcorgis

**Review by alfonssoo-sys — COMMENTED** [review]
283-2838256_windows-10-hero-wallpaper-black-windows-10-wallpaper

**Pull request merged**